File: Sources/Sentry/SentryFramesTracker.m
#import "SentryFramesTracker.h"

#if SENTRY_HAS_UIKIT

#    import "SentryCompiler.h"
#    import "SentryCurrentDateProvider.h"
#    import "SentryDelayedFrame.h"
#    import "SentryDisplayLinkWrapper.h"
#    import "SentryLog.h"
#    import "SentryProfiler.h"
#    import "SentryProfilingConditionals.h"
#    import "SentryTime.h"
#    import "SentryTracer.h"
#    import <SentryScreenFrames.h>
#    include <stdatomic.h>

#    if SENTRY_TARGET_PROFILING_SUPPORTED
/** A mutable version of @c SentryFrameInfoTimeSeries so we can accumulate results. */
typedef NSMutableArray<NSDictionary<NSString *, NSNumber *> *> SentryMutableFrameInfoTimeSeries;
#    endif // SENTRY_TARGET_PROFILING_SUPPORTED

static CFTimeInterval const SentryFrozenFrameThreshold = 0.7;
static CFTimeInterval const SentryPreviousFrameInitialValue = -1;

@interface
SentryFramesTracker ()

@property (nonatomic, strong, readonly) SentryDisplayLinkWrapper *displayLinkWrapper;
@property (nonatomic, strong, readonly) SentryCurrentDateProvider *dateProvider;
@property (nonatomic, assign) CFTimeInterval previousFrameTimestamp;
@property (nonatomic) uint64_t previousFrameSystemTimestamp;
@property (nonatomic) uint64_t currentFrameRate;
@property (nonatomic, strong) NSHashTable<id<SentryFramesTrackerListener>> *listeners;
#    if SENTRY_TARGET_PROFILING_SUPPORTED
@property (nonatomic, readwrite) SentryMutableFrameInfoTimeSeries *frozenFrameTimestamps;
@property (nonatomic, readwrite) SentryMutableFrameInfoTimeSeries *slowFrameTimestamps;
@property (nonatomic, readwrite) SentryMutableFrameInfoTimeSeries *frameRateTimestamps;
#    endif // SENTRY_TARGET_PROFILING_SUPPORTED

@property (nonatomic, strong) NSMutableArray<SentryDelayedFrame *> *delayedFrames;
@property (nonatomic) uint8_t delayedFrameWriteIndex;

@end

CFTimeInterval
slowFrameThreshold(uint64_t actualFramesPerSecond)
{
    // Most frames take just a few microseconds longer than the optimal calculated duration.
    // Therefore we subtract one, because otherwise almost all frames would be slow.
    return 1.0 / (actualFramesPerSecond - 1.0);
}

@implementation SentryFramesTracker {
    unsigned int _totalFrames;
    unsigned int _slowFrames;
    unsigned int _frozenFrames;
}

- (instancetype)initWithDisplayLinkWrapper:(SentryDisplayLinkWrapper *)displayLinkWrapper
                              dateProvider:(SentryCurrentDateProvider *)dateProvider;
{
    if (self = [super init]) {
        _isRunning = NO;
        _displayLinkWrapper = displayLinkWrapper;
        _dateProvider = dateProvider;
        _listeners = [NSHashTable weakObjectsHashTable];

        _currentFrameRate = 60;
        [self resetFrames];
        SENTRY_LOG_DEBUG(@"Initialized frame tracker %@", self);
    }
    return self;
}

/** Internal for testing */
- (void)setDisplayLinkWrapper:(SentryDisplayLinkWrapper *)displayLinkWrapper
{
    _displayLinkWrapper = displayLinkWrapper;
}

/** Internal for testing */
- (void)resetFrames
{
    _totalFrames = 0;
    _frozenFrames = 0;
    _slowFrames = 0;

    self.previousFrameTimestamp = SentryPreviousFrameInitialValue;
#    if SENTRY_TARGET_PROFILING_SUPPORTED
    [self resetProfilingTimestamps];
#    endif // SENTRY_TARGET_PROFILING_SUPPORTED

    [self resetDelayedFramesTimeStamps];
}

#    if SENTRY_TARGET_PROFILING_SUPPORTED
- (void)resetProfilingTimestamps
{
    self.frozenFrameTimestamps = [SentryMutableFrameInfoTimeSeries array];
    self.slowFrameTimestamps = [SentryMutableFrameInfoTimeSeries array];
    self.frameRateTimestamps = [SentryMutableFrameInfoTimeSeries array];
}
#    endif // SENTRY_TARGET_PROFILING_SUPPORTED

- (void)resetDelayedFramesTimeStamps
{
    _delayedFrames = [NSMutableArray array];
    SentryDelayedFrame *initialFrame =
        [[SentryDelayedFrame alloc] initWithStartTimestamp:[self.dateProvider systemTime]
                                          expectedDuration:0
                                            actualDuration:0];
    [_delayedFrames addObject:initialFrame];
    _delayedFrameWriteIndex = 1;
}

- (void)start
{
    if (_isRunning) {
        return;
    }

    _isRunning = YES;

    [_displayLinkWrapper linkWithTarget:self selector:@selector(displayLinkCallback)];
}

- (void)displayLinkCallback
{
    CFTimeInterval thisFrameTimestamp = self.displayLinkWrapper.timestamp;
    uint64_t thisFrameSystemTimestamp = self.dateProvider.systemTime;

    if (self.previousFrameTimestamp == SentryPreviousFrameInitialValue) {
        self.previousFrameTimestamp = thisFrameTimestamp;
        self.previousFrameSystemTimestamp = thisFrameSystemTimestamp;
        [self reportNewFrame];
        return;
    }

    // Calculate the actual frame rate as pointed out by the Apple docs:
    // https://developer.apple.com/documentation/quartzcore/cadisplaylink?language=objc The actual
    // frame rate can change at any time by setting preferredFramesPerSecond or due to ProMotion
    // display, low power mode, critical thermal state, and accessibility settings. Therefore we
    // need to check the frame rate for every callback.
    // targetTimestamp is only available on iOS 10.0 and tvOS 10.0 and above. We use a fallback of
    // 60 fps.
    _currentFrameRate = 60;
    if (UNLIKELY((self.displayLinkWrapper.targetTimestamp == self.displayLinkWrapper.timestamp))) {
        _currentFrameRate = 60;
    } else {
        _currentFrameRate = (uint64_t)round(
            (1 / (self.displayLinkWrapper.targetTimestamp - self.displayLinkWrapper.timestamp)));
    }

#    if SENTRY_TARGET_PROFILING_SUPPORTED
    if ([SentryProfiler isCurrentlyProfiling]) {
        BOOL hasNoFrameRatesYet = self.frameRateTimestamps.count == 0;
        uint64_t previousFrameRate
            = self.frameRateTimestamps.lastObject[@"value"].unsignedLongLongValue;
        BOOL frameRateChanged = previousFrameRate != _currentFrameRate;
        BOOL shouldRecordNewFrameRate = hasNoFrameRatesYet || frameRateChanged;
        if (shouldRecordNewFrameRate) {
            SENTRY_LOG_DEBUG(@"Recording new frame rate at %llu.", thisFrameSystemTimestamp);
            [self recordTimestamp:thisFrameSystemTimestamp
                            value:@(_currentFrameRate)
                            array:self.frameRateTimestamps];
        }
    }
#    endif // SENTRY_TARGET_PROFILING_SUPPORTED

    CFTimeInterval frameDuration = thisFrameTimestamp - self.previousFrameTimestamp;

    if (frameDuration > slowFrameThreshold(_currentFrameRate)
        && frameDuration <= SentryFrozenFrameThreshold) {
        _slowFrames++;
#    if SENTRY_TARGET_PROFILING_SUPPORTED
        SENTRY_LOG_DEBUG(@"Capturing slow frame starting at %llu (frame tracker: %@).",
            thisFrameSystemTimestamp, self);
        [self recordTimestamp:thisFrameSystemTimestamp
                        value:@(thisFrameSystemTimestamp - self.previousFrameSystemTimestamp)
                        array:self.slowFrameTimestamps];
#    endif // SENTRY_TARGET_PROFILING_SUPPORTED
    } else if (frameDuration > SentryFrozenFrameThreshold) {
        _frozenFrames++;
#    if SENTRY_TARGET_PROFILING_SUPPORTED
        SENTRY_LOG_DEBUG(@"Capturing frozen frame starting at %llu.", thisFrameSystemTimestamp);
        [self recordTimestamp:thisFrameSystemTimestamp
                        value:@(thisFrameSystemTimestamp - self.previousFrameSystemTimestamp)
                        array:self.frozenFrameTimestamps];
#    endif // SENTRY_TARGET_PROFILING_SUPPORTED
    }

    if (frameDuration > slowFrameThreshold(_currentFrameRate)) {
        [self recordDelayedFrame:self.previousFrameSystemTimestamp
                expectedDuration:slowFrameThreshold(_currentFrameRate)
                  actualDuration:frameDuration];
    }

    _totalFrames++;
    self.previousFrameTimestamp = thisFrameTimestamp;
    self.previousFrameSystemTimestamp = thisFrameSystemTimestamp;
    [self reportNewFrame];
}

- (void)reportNewFrame
{
    NSArray *localListeners;
    @synchronized(self.listeners) {
        localListeners = [self.listeners allObjects];
    }

    for (id<SentryFramesTrackerListener> listener in localListeners) {
        [listener framesTrackerHasNewFrame];
    }
}

#    if SENTRY_TARGET_PROFILING_SUPPORTED
- (void)recordTimestamp:(uint64_t)timestamp value:(NSNumber *)value array:(NSMutableArray *)array
{
    BOOL shouldRecord = [SentryProfiler isCurrentlyProfiling];
#        if defined(TEST) || defined(TESTCI)
    shouldRecord = YES;
#        endif // defined(TEST) || defined(TESTCI)
    if (shouldRecord) {
        [array addObject:@{ @"timestamp" : @(timestamp), @"value" : value }];
    }
}
#    endif // SENTRY_TARGET_PROFILING_SUPPORTED

- (SentryScreenFrames *)currentFrames
{
#    if SENTRY_TARGET_PROFILING_SUPPORTED
    return [[SentryScreenFrames alloc] initWithTotal:_totalFrames
                                              frozen:_frozenFrames
                                                slow:_slowFrames
                                 slowFrameTimestamps:self.slowFrameTimestamps
                               frozenFrameTimestamps:self.frozenFrameTimestamps
                                 frameRateTimestamps:self.frameRateTimestamps];
#    else
    return [[SentryScreenFrames alloc] initWithTotal:_totalFrames
                                              frozen:_frozenFrames
                                                slow:_slowFrames];
#    endif // SENTRY_TARGET_PROFILING_SUPPORTED
}

- (void)recordDelayedFrame:(uint64_t)startSystemTimestamp
          expectedDuration:(CFTimeInterval)expectedDuration
            actualDuration:(CFTimeInterval)actualDuration
{
    @synchronized(self.delayedFrames) {
        [self removeOldDelayedFrames];

        SENTRY_LOG_DEBUG(@"FrameDelay: Record expected:%f ms actual %f ms", expectedDuration * 1000,
            actualDuration * 1000);
        SentryDelayedFrame *delayedFrame =
            [[SentryDelayedFrame alloc] initWithStartTimestamp:startSystemTimestamp
                                              expectedDuration:expectedDuration
                                                actualDuration:actualDuration];
        [self.delayedFrames addObject:delayedFrame];
    }
}

/**
 * Removes delayed frame that are older than current time minus
 * SENTRY_AUTO_TRANSACTION_MAX_DURATION.
 * @note Make sure to call this in a @synchronized block.
 */
- (void)removeOldDelayedFrames
{
    u_int64_t transactionMaxDurationNS
        = timeIntervalToNanoseconds(SENTRY_AUTO_TRANSACTION_MAX_DURATION);

    uint64_t removeFramesBeforeSystemTimeStamp
        = _dateProvider.systemTime - transactionMaxDurationNS;
    if (_dateProvider.systemTime < transactionMaxDurationNS) {
        removeFramesBeforeSystemTimeStamp = 0;
    }

    NSInteger i = 0;
    for (SentryDelayedFrame *frame in self.delayedFrames) {
        uint64_t frameEndSystemTimeStamp
            = frame.startSystemTimestamp + timeIntervalToNanoseconds(frame.actualDuration);
        if (frameEndSystemTimeStamp < removeFramesBeforeSystemTimeStamp) {
            i++;
        } else {
            break;
        }
    }
    [self.delayedFrames removeObjectsInRange:NSMakeRange(0, i)];
}

- (CFTimeInterval)getFrameDelay:(uint64_t)startSystemTimestamp
             endSystemTimestamp:(uint64_t)endSystemTimestamp
{
    CFTimeInterval cantCalculateFrameDelay = -1.0;

    if (_isRunning == NO) {
        SENTRY_LOG_DEBUG(@"Not calculating frames delay because frames tracker isn't running.");
        return cantCalculateFrameDelay;
    }

    if (startSystemTimestamp >= endSystemTimestamp) {
        SENTRY_LOG_DEBUG(@"Not calculating frames delay because startSystemTimestamp is before  "
                         @"endSystemTimestamp");
        return cantCalculateFrameDelay;
    }

    if (endSystemTimestamp > self.dateProvider.systemTime) {
        SENTRY_LOG_DEBUG(@"Not calculating frames delay endSystemTimestamp is in the future.");
        return cantCalculateFrameDelay;
    }

    // Check if there is an delayed frame going on but not recorded yet.
    CFTimeInterval thisFrameTimestamp = self.displayLinkWrapper.timestamp;
    CFTimeInterval frameDuration = thisFrameTimestamp - self.previousFrameTimestamp;
    CFTimeInterval ongoingDelayedFrame = 0.0;
    if (frameDuration > slowFrameThreshold(_currentFrameRate)) {
        ongoingDelayedFrame = frameDuration - slowFrameThreshold(_currentFrameRate);
    }

    CFTimeInterval delay = ongoingDelayedFrame;

    @synchronized(self.delayedFrames) {

        // Although we use a ring buffer, and it could take a while until we get to frames within
        // the time interval for which we want to return frame delay, we don't want to use a normal
        // for loop with accessing the elements by index because NSFastEnumeration is faster than a
        // normal for loop and in the worst case we need to iterate over all elements anyways. A
        // binary search to find the first and last delayed frame in our time interval doesn't help
        // either because, again, we need to iterate over all elements in the worst case.

        uint64_t oldestDelayedFrameStartTimestamp = UINT64_MAX;

        for (SentryDelayedFrame *frame in self.delayedFrames) {
            if (frame.startSystemTimestamp < oldestDelayedFrameStartTimestamp) {
                oldestDelayedFrameStartTimestamp = frame.startSystemTimestamp;
            }
        }

        if (oldestDelayedFrameStartTimestamp > startSystemTimestamp) {
            return cantCalculateFrameDelay;
        }

        for (SentryDelayedFrame *frame in self.delayedFrames) {
            NSDate *startDate =
                [NSDate dateWithTimeIntervalSinceReferenceDate:nanosecondsToTimeInterval(
                                                                   startSystemTimestamp)];
            NSDate *endDate =
                [NSDate dateWithTimeIntervalSinceReferenceDate:nanosecondsToTimeInterval(
                                                                   endSystemTimestamp)];

            CFTimeInterval delayStartTime
                = nanosecondsToTimeInterval(frame.startSystemTimestamp) + frame.expectedDuration;
            NSDate *frameDelayStartDate =
                [NSDate dateWithTimeIntervalSinceReferenceDate:delayStartTime];

            NSDateInterval *frameDelayDateInterval =
                [[NSDateInterval alloc] initWithStartDate:startDate endDate:endDate];

            // Only use the date interval for the actual delay
            NSDateInterval *frameDateInterval = [[NSDateInterval alloc]
                initWithStartDate:frameDelayStartDate
                         duration:(frame.actualDuration - frame.expectedDuration)];

            if ([frameDelayDateInterval intersectsDateInterval:frameDateInterval]) {
                NSDateInterval *intersection =
                    [frameDelayDateInterval intersectionWithDateInterval:frameDateInterval];
                delay = delay + intersection.duration;
            }
        }
    }
    return delay;
}

- (void)addListener:(id<SentryFramesTrackerListener>)listener
{

    @synchronized(self.listeners) {
        [self.listeners addObject:listener];
    }
}

- (void)removeListener:(id<SentryFramesTrackerListener>)listener
{
    @synchronized(self.listeners) {
        [self.listeners removeObject:listener];
    }
}

- (void)stop
{
    _isRunning = NO;
    [self.displayLinkWrapper invalidate];
    [self resetDelayedFramesTimeStamps];
}

- (void)dealloc
{
    [self stop];
}

@end

BOOL
sentryShouldAddSlowFrozenFramesData(
    NSInteger totalFrames, NSInteger slowFrames, NSInteger frozenFrames)
{
    BOOL allBiggerThanZero = totalFrames >= 0 && slowFrames >= 0 && frozenFrames >= 0;
    BOOL oneBiggerThanZero = totalFrames > 0 || slowFrames > 0 || frozenFrames > 0;

    return allBiggerThanZero && oneBiggerThanZero;
}

#endif // SENTRY_HAS_UIKIT

File: Sources/Swift/Core/Integrations/FramesTracking/SentryFramesTracker.swift
// swiftlint:disable file_length type_body_length
@_implementationOnly import _SentryPrivate

#if (os(iOS) || os(tvOS) || (swift(>=5.9) && os(visionOS))) && !SENTRY_NO_UIKIT
import UIKit
private typealias CrossPlatformApplication = UIApplication
#elseif os(macOS)
import AppKit
private typealias CrossPlatformApplication = NSApplication
#endif

#if (os(iOS) || os(tvOS) || (swift(>=5.9) && os(visionOS))) && !SENTRY_NO_UIKIT

@_spi(Private) @objc
public protocol SentryFramesTrackerListener: NSObjectProtocol {
    func framesTrackerHasNewFrame(_ newFrameDate: Date)
}

@_spi(Private) @objc
public class SentryFramesTracker: NSObject {

    var isStarted: Bool = false
    @objc public private(set) var isRunning: Bool = false
    
    // MARK: Private properties
    private var lock = NSLock()
    private var previousFrameTimestamp: CFTimeInterval = SentryFramesTracker.previousFrameInitialValue
    private var previousFrameSystemTimestamp: UInt64 = 0
    private var currentFrameRate: UInt64 = 60
    private var listeners: NSHashTable<SentryFramesTrackerListener>

#if os(iOS)
    private var frozenFrameTimestamps = SentryFrameInfoTimeSeries()
    private var slowFrameTimestamps = SentryFrameInfoTimeSeries()
    private var frameRateTimestamps = SentryFrameInfoTimeSeries()
#endif // os(iOS)

    // Frame counters - accessed only from main thread (display link callback)
    private var totalFrames: UInt = 0
    private var slowFrames: UInt = 0
    private var frozenFrames: UInt = 0

    private var displayLinkWrapper: SentryDisplayLinkWrapper
    private let dateProvider: SentryCurrentDateProvider
    private let dispatchQueueWrapper: SentryDispatchQueueWrapper
    private let notificationCenter: SentryNSNotificationCenterWrapper
    private var delayedFramesTracker: SentryDelayedFramesTrackerWrapper
    
    private static let frozenFrameThreshold: CFTimeInterval = 0.7
    private static let previousFrameInitialValue: CFTimeInterval = -1

    init(
        displayLinkWrapper: SentryDisplayLinkWrapper,
        dateProvider: SentryCurrentDateProvider,
        dispatchQueueWrapper: SentryDispatchQueueWrapper,
        notificationCenter: SentryNSNotificationCenterWrapper,
        delayedFramesTracker: SentryDelayedFramesTrackerWrapper
    ) {
        self.displayLinkWrapper = displayLinkWrapper
        self.dateProvider = dateProvider
        self.dispatchQueueWrapper = dispatchQueueWrapper
        self.notificationCenter = notificationCenter
        self.delayedFramesTracker = delayedFramesTracker
        self.listeners = NSHashTable<SentryFramesTrackerListener>.weakObjects()

        super.init()

        resetFrames()
        SentrySDKLog.debug("Initialized frame tracker")
    }

    // MARK: - Public Methods

    @objc
    public func start() {
        guard !isStarted else { return }

        isStarted = true

        notificationCenter.addObserver(
            self,
            selector: #selector(didBecomeActive),
            name: CrossPlatformApplication.didBecomeActiveNotification,
            object: nil
        )

        notificationCenter.addObserver(
            self,
            selector: #selector(willResignActive),
            name: CrossPlatformApplication.willResignActiveNotification,
            object: nil
        )

        unpause()
    }

    @objc
    public func stop() {
        guard isStarted else { return }

        isStarted = false

        pause()

        notificationCenter.removeObserver(
            self,
            name: CrossPlatformApplication.didBecomeActiveNotification,
            object: nil
        )

        notificationCenter.removeObserver(
            self,
            name: CrossPlatformApplication.willResignActiveNotification,
            object: nil
        )

        lock.synchronized {
            listeners.removeAllObjects()
        }
    }

    @objc
    public func currentFrames() -> SentryScreenFrames {
#if os(iOS)
        return SentryScreenFrames(
            total: totalFrames,
            frozen: frozenFrames,
            slow: slowFrames,
            slowFrameTimestamps: slowFrameTimestamps,
            frozenFrameTimestamps: frozenFrameTimestamps,
            frameRateTimestamps: frameRateTimestamps
        )
#else
        return SentryScreenFrames(
            total: totalFrames,
            frozen: frozenFrames,
            slow: slowFrames
        )
#endif
    }

    func getFramesDelay(
        _ startSystemTimestamp: UInt64,
        endSystemTimestamp: UInt64
    ) -> SentryFramesDelayResult {
        return delayedFramesTracker.getFramesDelay(
            startSystemTimestamp,
            endSystemTimestamp: endSystemTimestamp,
            isRunning: isRunning,
            slowFrameThreshold: Self.slowFrameThreshold(currentFrameRate)
        )
    }
    
    @objc public func getFramesDelaySPI(
        _ startSystemTimestamp: UInt64,
        endSystemTimestamp: UInt64
    ) -> SentryFramesDelayResultSPI {
        let result = getFramesDelay(startSystemTimestamp, endSystemTimestamp: endSystemTimestamp)
        return .init(delayDuration: result.delayDuration, framesContributingToDelayCount: result.framesContributingToDelayCount)
    }

    @objc public func addListener(_ listener: SentryFramesTrackerListener) {
        dispatchQueueWrapper.dispatchAsyncOnMainQueueIfNotMainThread {
            self.listeners.add(listener)
        }
    }

    @objc public func removeListener(_ listener: SentryFramesTrackerListener) {
        dispatchQueueWrapper.dispatchAsyncOnMainQueueIfNotMainThread {
            self.listeners.remove(listener)
        }
    }
    
    deinit {
        stop()
    }

#if os(iOS)
    @objc public func resetProfilingTimestamps() {
        // The DisplayLink callback always runs on the main thread. We dispatch this to the main thread
        // instead to avoid using locks in the DisplayLink callback.
        dispatchQueueWrapper.dispatchAsyncOnMainQueueIfNotMainThread {
            self.resetProfilingTimestampsInternal()
        }
    }
#endif // os(iOS)

    // MARK: - Internal Methods

#if SENTRY_TEST || SENTRY_TEST_CI
    @objc func setDisplayLinkWrapper(_ displayLinkWrapper: SentryDisplayLinkWrapper) {
        self.displayLinkWrapper = displayLinkWrapper
    }
    
    var listenersCount: Int {
        listeners.count
    }
#endif

    @objc func resetFrames() {
        totalFrames = 0
        frozenFrames = 0
        slowFrames = 0

        previousFrameTimestamp = Self.previousFrameInitialValue

#if os(iOS)
        resetProfilingTimestampsInternal()
#endif

        delayedFramesTracker.reset()
    }

    // MARK: - Private Methods

    @objc
    private func didBecomeActive() {
        unpause()
    }

    @objc
    private func willResignActive() {
        pause()
    }

    private func unpause() {
        guard !isRunning else { return }

        isRunning = true

        // Reset the previous frame timestamp to avoid wrong metrics being collected
        previousFrameTimestamp = Self.previousFrameInitialValue
        displayLinkWrapper.link(withTarget: self, selector: #selector(displayLinkCallback))
    }

    private func pause() {
        isRunning = false

        // When the frames tracker is paused, we must reset the delayed frames tracker since accurate
        // frame delay statistics cannot be provided, as we can't account for events during the pause.
        delayedFramesTracker.reset()

        displayLinkWrapper.invalidate()
    }

// swiftlint:disable file_length function_body_length
    @objc
    private func displayLinkCallback() {
        let thisFrameTimestamp = displayLinkWrapper.timestamp
        let thisFrameSystemTimestamp = dateProvider.systemTime()

        if previousFrameTimestamp == Self.previousFrameInitialValue {
            previousFrameTimestamp = thisFrameTimestamp
            previousFrameSystemTimestamp = thisFrameSystemTimestamp
            delayedFramesTracker.setPreviousFrameSystemTimestamp(thisFrameSystemTimestamp)
            reportNewFrame()
            return
        }

        // Calculate the actual frame rate as pointed out by the Apple docs:
        // https://developer.apple.com/documentation/quartzcore/cadisplaylink?language=objc The actual
        // frame rate can change at any time by setting preferredFramesPerSecond or due to ProMotion
        // display, low power mode, critical thermal state, and accessibility settings. Therefore we
        // need to check the frame rate for every callback.
        // targetTimestamp is only available on iOS 10.0 and tvOS 10.0 and above. We use a fallback of
        // 60 fps.
        currentFrameRate = 60
        if displayLinkWrapper.targetTimestamp != displayLinkWrapper.timestamp {
            currentFrameRate = UInt64(round(
                1 / (displayLinkWrapper.targetTimestamp - displayLinkWrapper.timestamp)
            ))
        }

#if os(iOS)
        let thisFrameNSDate = dateProvider.date()
        let isContinuousProfiling = SentryContinuousProfiler.isCurrentlyProfiling()
        let profilingTimestamp = isContinuousProfiling ?
            NSNumber(value: thisFrameNSDate.timeIntervalSince1970) :
            NSNumber(value: thisFrameSystemTimestamp)

        if SentryTraceProfiler.isCurrentlyProfiling() || isContinuousProfiling {
            let hasNoFrameRatesYet = frameRateTimestamps.isEmpty
            let previousFrameRate = frameRateTimestamps.last?["value"]?.uint64Value ?? 0
            let frameRateChanged = previousFrameRate != currentFrameRate
            let shouldRecordNewFrameRate = hasNoFrameRatesYet || frameRateChanged

            if shouldRecordNewFrameRate {
                SentrySDKLog.debug("Recording new frame rate at \(profilingTimestamp).")
                recordTimestamp(profilingTimestamp, value: NSNumber(value: currentFrameRate), array: &frameRateTimestamps)
            }
        }
#endif // os(iOS)

        let frameDuration = thisFrameTimestamp - previousFrameTimestamp
        let slowThreshold = Self.slowFrameThreshold(currentFrameRate)

        if frameDuration > slowThreshold && frameDuration <= Self.frozenFrameThreshold {
            slowFrames += 1
#if os(iOS)
            SentrySDKLog.debug("Detected slow frame starting at \(profilingTimestamp) (frame tracker: \(self)).")
            recordTimestamp(
                profilingTimestamp,
                value: NSNumber(value: thisFrameSystemTimestamp - previousFrameSystemTimestamp),
                array: &slowFrameTimestamps
            )
#endif // os(iOS)
        } else if frameDuration > Self.frozenFrameThreshold {
            frozenFrames += 1
#if os(iOS)
            SentrySDKLog.debug("Detected frozen frame starting at \(profilingTimestamp).")
            recordTimestamp(
                profilingTimestamp,
                value: NSNumber(value: thisFrameSystemTimestamp - previousFrameSystemTimestamp),
                array: &frozenFrameTimestamps
            )
#endif // os(iOS)
        }

        if frameDuration > slowThreshold {
            delayedFramesTracker.recordDelayedFrame(
                previousFrameSystemTimestamp,
                thisFrameSystemTimestamp: thisFrameSystemTimestamp,
                expectedDuration: slowThreshold,
                actualDuration: frameDuration
            )
        } else {
            delayedFramesTracker.setPreviousFrameSystemTimestamp(thisFrameSystemTimestamp)
        }

        totalFrames += 1
        previousFrameTimestamp = thisFrameTimestamp
        previousFrameSystemTimestamp = thisFrameSystemTimestamp
        reportNewFrame()
    }
// swiftlint:enable file_length function_body_length

    private func reportNewFrame() {
        let newFrameDate = dateProvider.date()
        // We need to copy the list because some listeners will remove themselves
        // from the list during the callback, causing a crash during iteration.
        let listenersArray = self.listeners.allObjects
        for listener in listenersArray {
            listener.framesTrackerHasNewFrame(newFrameDate)
        }
    }

#if os(iOS)
    private func recordTimestamp(_ timestamp: NSNumber, value: NSNumber, array: inout SentryFrameInfoTimeSeries) {
        var shouldRecord = SentryTraceProfiler.isCurrentlyProfiling() || SentryContinuousProfiler.isCurrentlyProfiling()
#if SENTRY_TEST || SENTRY_TEST_CI
        shouldRecord = true
#endif
        if shouldRecord {
            array.append(["timestamp": timestamp, "value": value])
        }
    }
    
    private func resetProfilingTimestampsInternal() {
        SentrySDKLog.debug("Resetting profiling GPU timeseries data.")
        frozenFrameTimestamps = []
        slowFrameTimestamps = []
        frameRateTimestamps = []
    }
#endif // os(iOS)
    
    // MARK: - Static Functions
    @objc
    public static func shouldAddSlowFrozenFramesData(
        totalFrames: Int,
        slowFrames: Int,
        frozenFrames: Int
    ) -> Bool {
        let allBiggerThanOrEqualToZero = totalFrames >= 0 && slowFrames >= 0 && frozenFrames >= 0
        let oneBiggerThanZero = totalFrames > 0 || slowFrames > 0 || frozenFrames > 0

        return allBiggerThanOrEqualToZero && oneBiggerThanZero
    }
    
    static func slowFrameThreshold(_ actualFramesPerSecond: UInt64) -> CFTimeInterval {
        // Most frames take just a few microseconds longer than the optimal calculated duration.
        // Therefore we subtract one, because otherwise almost all frames would be slow.
        return 1.0 / CFTimeInterval(actualFramesPerSecond - 1)
    }
}

#endif
// swiftlint:enable file_length type_body_length
